packages/opentelemetry-plugin-http/src/utils.ts

```diff
@@ -0,0 +1,214 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Status, CanonicalCode, Span } from '@opentelemetry/types';
+import {
+  RequestOptions,
+  IncomingMessage,
+  ClientRequest,
+  IncomingHttpHeaders,
+} from 'http';
+import { IgnoreMatcher, ParsedRequestOptions } from './types';
+import { AttributeNames } from './enums/attributeNames';
+import * as url from 'url';
+
+/**
+ * Utility class
+ */
+export class Utils {
+  /**
+   * return an absolute url
+   */
+  static getUrlFromIncomingRequest(
+    requestUrl: url.UrlWithStringQuery | null,
+    headers: IncomingHttpHeaders
+  ): string {
+    if (!requestUrl) {
+      return `http://${headers.host || 'localhost'}/`;
+    }
+
+    return requestUrl.href && requestUrl.href.startsWith('http')
+      ? `${requestUrl.protocol}//${requestUrl.hostname}${requestUrl.path}`
+      : `${requestUrl.protocol || 'http:'}//${headers.host ||
+          'localhost'}${requestUrl.path || '/'}`;
+  }
+  /**
+   * Parse status code from HTTP response.
+   */
+  static parseResponseStatus(statusCode: number): Omit<Status, 'message'> {
+    if (statusCode < 200 || statusCode > 504) {
+      return { code: CanonicalCode.UNKNOWN };
+    } else if (statusCode >= 200 && statusCode < 400) {
+      return { code: CanonicalCode.OK };
+    } else {
+      switch (statusCode) {
+        case 400:
+          return { code: CanonicalCode.INVALID_ARGUMENT };
+        case 504:
+          return { code: CanonicalCode.DEADLINE_EXCEEDED };
+        case 404:
+          return { code: CanonicalCode.NOT_FOUND };
+        case 403:
+          return { code: CanonicalCode.PERMISSION_DENIED };
+        case 401:
+          return { code: CanonicalCode.UNAUTHENTICATED };
+        case 429:
+          return { code: CanonicalCode.RESOURCE_EXHAUSTED };
+        case 501:
+          return { code: CanonicalCode.UNIMPLEMENTED };
+        case 503:
+          return { code: CanonicalCode.UNAVAILABLE };
+        default:
+          return { code: CanonicalCode.UNKNOWN };
+      }
+    }
+  }
+
+  /**
+   * Returns whether the Expect header is on the given options object.
+   * @param options Options for http.request.
+   */
+  static hasExpectHeader(options: RequestOptions | url.URL): boolean {
+    return !!(
+      (options as RequestOptions).headers &&
+      (options as RequestOptions).headers!.Expect
```

Comment:
I think this one is used during parsing headers of ClientRequests where user code defines the case used for header names and as a result using `Expect` with an uppercase `E` may not match in any case.